lib/src/rules/avoid_types_on_closure_parameters.dart

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.avoid_annotating_types_on_function_expressions;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/analyzer.dart';
+import 'package:linter/src/util/dart_type_utilities.dart';
+
+const _desc = r'Avoid annotating types on function expressions.';
+
+const _details = r'''
+
+**AVOID** annotating types on function expressions.
+
+**BAD:**
+```
+var names = people.map((Person person) => person.name);
+```
+
+**GOOD:**
+```
+var names = people.map((person) => person.name);
+```
+
+''';
+
+bool _doesNotHaveType(SimpleFormalParameter parameter) {
+  return parameter?.type != null && parameter.type.name.name != 'dynamic';
+}
+
+SimpleFormalParameter _getSimpleFormalParameter(FormalParameter node) {
+  if (node is SimpleFormalParameter) {
+    return node;
+  }
+  final simpleFormalParameters = DartTypeUtilities
+      .traverseNodesInDFS(node)
+      .where(_isSimpleFormalParameter);
+  return simpleFormalParameters.isNotEmpty
```

Comment:
firstWhere returns an Exception if does not found, and that happens with FunctionTypedFormalParameter